File: frontend/src/lib/streak.ts
// Streak day-scoring — mirrors the backend in luma/services/streak.py. The Today
// flame (StreakStrip) and the per-day history (StreakHistorySheet) both grade days
// through here so they can never disagree with each other or with the server.
//
// Calories use an asymmetric band: a day still counts when intake is UNDER target
// (the point of a deficit) down to 25% below, but only 10% above. Saturated fat and
// sodium are ceilings (≤ target); soluble fiber is a floor (≥ target). A target the
// user hasn't set is 'untracked' — excluded from the tally, never counted as a miss.

export const CAL_UNDER_TOL = 0.75
export const CAL_OVER_TOL = 1.1
export const CEILING_GRACE = 1.10   // sat fat, sodium: up to 10% over cap still counts
export const FLOOR_TOLERANCE = 0.90 // soluble fiber: down to 10% under floor still counts
export const ON_TRACK_MIN = 3

export type MetricState = 'met' | 'missed' | 'untracked'

type Num = number | null | undefined

export function calorieState(logged: Num, target: Num): MetricState {
  if (target == null) return 'untracked'
  const l = logged ?? 0
  return l >= target * CAL_UNDER_TOL && l <= target * CAL_OVER_TOL ? 'met' : 'missed'
}

/** Ceiling metric (saturated fat, sodium): met within 10% grace over target. */
export function ceilingState(logged: Num, target: Num): MetricState {
  if (target == null) return 'untracked'
  return (logged ?? 0) <= target * CEILING_GRACE ? 'met' : 'missed'
}

/** Floor metric (soluble fiber): met within 10% grace under target. */
export function floorState(logged: Num, target: Num): MetricState {
  if (target == null) return 'untracked'
  return (logged ?? 0) >= target * FLOOR_TOLERANCE ? 'met' : 'missed'
}

export interface DayInputs {
  cal?: Num
  sat?: Num
  fib?: Num
  sod?: Num
}

export interface DayScore {
  cal: MetricState
  sat: MetricState
  fib: MetricState
  sod: MetricState
  targetsMet: number
  targetsPossible: number
  onTrack: boolean
}

export function scoreDay(logged: DayInputs, targets: DayInputs): DayScore {
  const cal = calorieState(logged.cal, targets.cal)
  const sat = ceilingState(logged.sat, targets.sat)
  const fib = floorState(logged.fib, targets.fib)
  const sod = ceilingState(logged.sod, targets.sod)
  const states = [cal, sat, fib, sod]
  const targetsPossible = states.filter(s => s !== 'untracked').length
  const targetsMet = states.filter(s => s === 'met').length
  const onTrack = targetsPossible > 0 && targetsMet >= Math.min(ON_TRACK_MIN, targetsPossible)
  return { cal, sat, fib, sod, targetsMet, targetsPossible, onTrack }
}

File: backend/luma/services/streak.py
"""Shared scoring for the Today streak flame and the streak-history breakdown.

The headline streak in ``/today`` and the per-day list in ``/today/streak-history``
MUST grade a day identically, otherwise the big number contradicts the breakdown
(the original bug: the headline counted any day a meal was logged while the list
counted days that hit the targets). Both now call :func:`score_day`. The frontend
mirrors this exact logic in ``src/lib/streak.ts`` — keep the two in sync.

Calorie scoring is intentionally asymmetric. A day still counts when intake runs
UNDER target (the whole point of a deficit) down to 15% below, but only 10% above,
because overshooting calories is what breaks a cut — landing under it does not.
Saturated fat and sugar are ceilings (at or below target); soluble fiber is a floor
(at or above target). A target the user has not set is excluded from the tally
entirely rather than scored as a miss, so an unconfigured metric can never sink a
streak.
"""
from __future__ import annotations

CAL_UNDER_TOL = 0.85   # intake down to 15% under target is still on-plan
CAL_OVER_TOL = 1.10    # only 10% over target is tolerated
ON_TRACK_MIN = 3       # hit at least this many targets — or all of them, if fewer are set


def score_day(
    totals: dict[str, float],
    targets: dict[str, float | None],
) -> dict[str, object]:
    """Grade one day's nutrient totals against the configured targets.

    ``totals`` and ``targets`` are keyed by ``"cal"``, ``"sat"``, ``"fib"``,
    ``"sug"``. Only targets that are set (non-``None``) count toward the
    denominator. Returns per-metric ``*_met`` flags plus ``targets_met`` (numerator),
    ``targets_possible`` (denominator) and the ``on_track`` verdict.
    """
    checks: dict[str, bool] = {}

    cal_t = targets.get("cal")
    if cal_t is not None:
        checks["cal"] = cal_t * CAL_UNDER_TOL <= totals.get("cal", 0.0) <= cal_t * CAL_OVER_TOL

    sat_t = targets.get("sat")
    if sat_t is not None:
        checks["sat"] = totals.get("sat", 0.0) <= sat_t

    fib_t = targets.get("fib")
    if fib_t is not None:
        checks["fib"] = totals.get("fib", 0.0) >= fib_t

    sug_t = targets.get("sug")
    if sug_t is not None:
        checks["sug"] = totals.get("sug", 0.0) <= sug_t

    configured = len(checks)
    met = sum(checks.values())
    on_track = configured > 0 and met >= min(ON_TRACK_MIN, configured)

    return {
        "cal_met": checks.get("cal", False),
        "sat_met": checks.get("sat", False),
        "fib_met": checks.get("fib", False),
        "sug_met": checks.get("sug", False),
        "targets_met": met,
        "targets_possible": configured,
        "on_track": on_track,
    }
